File: ee/engine/license/license_test.go
package license_test

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"github.com/antifailure/antifailure/ee/engine/license"
)

var epoch = time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)

type signer struct {
	keyID string
	priv  ed25519.PrivateKey
	pub   ed25519.PublicKey
}

func newSigner(t *testing.T, keyID string) signer {
	t.Helper()
	pub, priv, err := ed25519.GenerateKey(nil)
	require.NoError(t, err)
	return signer{keyID: keyID, priv: priv, pub: pub}
}

// sign builds a token the way licensegen does, so the tests exercise the real
// wire form rather than a convenient one.
func (s signer) sign(t *testing.T, claims license.Claims) string {
	t.Helper()
	if claims.KeyID == "" {
		claims.KeyID = s.keyID
	}
	payload, err := json.Marshal(claims)
	require.NoError(t, err)
	return fmt.Sprintf("aflic_%s.%s",
		base64.RawURLEncoding.EncodeToString(payload),
		base64.RawURLEncoding.EncodeToString(ed25519.Sign(s.priv, payload)))
}

func validClaims() license.Claims {
	return license.Claims{
		ID:        "lic-1",
		Org:       "acme",
		Plan:      "enterprise",
		Features:  []license.Feature{license.FeatureSSO, license.FeatureSCIM},
		Seats:     50,
		IssuedAt:  epoch,
		ExpiresAt: epoch.AddDate(1, 0, 0),
	}
}

// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

func TestParse_AcceptsAKeyItSigned(t *testing.T) {
	t.Parallel()
	s := newSigner(t, "2026-01")
	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})

	claims, err := v.Parse(s.sign(t, validClaims()))
	require.NoError(t, err)
	require.Equal(t, "acme", claims.Org)
	require.Equal(t, 50, claims.Seats)
	require.Contains(t, claims.Features, license.FeatureSSO)
}

func TestParse_ToleratesWhitespaceFromAPastedKey(t *testing.T) {
	t.Parallel()
	// A license arrives in an email and gets pasted with a line break in it.
	// Refusing that is a support ticket, not a security property.
	s := newSigner(t, "k")
	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})
	token := s.sign(t, validClaims())

	middle := len(token) / 2
	wrapped := "  " + token[:middle] + "\n   " + token[middle:] + "\n"
	_, err := v.Parse(wrapped)
	require.NoError(t, err)
}

func TestParse_RefusesATamperedPayload(t *testing.T) {
	t.Parallel()
	s := newSigner(t, "k")
	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})

	// Every field somebody would want to change: more seats, a later expiry, a
	// different organization, an extra feature.
	for name, edit := range map[string]func(*license.Claims){
		"more seats":       func(c *license.Claims) { c.Seats = 100000 },
		"a later expiry":   func(c *license.Claims) { c.ExpiresAt = epoch.AddDate(50, 0, 0) },
		"another org":      func(c *license.Claims) { c.Org = "someone-else" },
		"an extra feature": func(c *license.Claims) { c.Features = license.AllFeatures() },
	} {
		t.Run(name, func(t *testing.T) {
			honest := validClaims()
			token := s.sign(t, honest)

			forged := validClaims()
			forged.KeyID = s.keyID
			edit(&forged)
			// The key id is set before marshalling, not after. Setting it after
			// leaves the forged payload naming no key, and the parse is then
			// refused for an unknown key rather than a bad signature, which
			// would pass this test while proving nothing about the signature.
			payload, err := json.Marshal(forged)
			require.NoError(t, err)

			// The original signature over the edited payload, which is what an
			// attacker with a real license can construct.
			parts := strings.SplitN(strings.TrimPrefix(token, "aflic_"), ".", 2)
			tampered := "aflic_" + base64.RawURLEncoding.EncodeToString(payload) + "." + parts[1]

			_, err = v.Parse(tampered)
			require.ErrorIs(t, err, license.ErrTampered)
		})
	}
}

func TestParse_RefusesAKeyItDoesNotKnow(t *testing.T) {
	t.Parallel()
	mine := newSigner(t, "mine")
	theirs := newSigner(t, "theirs")
	v := license.NewVerifier(map[string]ed25519.PublicKey{mine.keyID: mine.pub})

	_, err := v.Parse(theirs.sign(t, validClaims()))
	require.ErrorIs(t, err, license.ErrUnknownKey)
}

func TestParse_AcceptsAnOldKeyAfterRotation(t *testing.T) {
	t.Parallel()
	// A build that trusts only the newest key cannot verify any license already
	// in the field, which turns a key rotation into an outage for every
	// customer at once.
	old := newSigner(t, "2025-01")
	current := newSigner(t, "2026-01")
	v := license.NewVerifier(map[string]ed25519.PublicKey{
		old.keyID: old.pub, current.keyID: current.pub,
	})

	for _, s := range []signer{old, current} {
		_, err := v.Parse(s.sign(t, validClaims()))
		require.NoErrorf(t, err, "key %s", s.keyID)
	}
}

func TestParse_RefusesMalformedInputWithoutPanicking(t *testing.T) {
	t.Parallel()
	s := newSigner(t, "k")
	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})

	for _, token := range []string{
		"", "   ", "not-a-license", "aflic_", "aflic_.", "aflic_a", "aflic_a.",
		"aflic_.b", "aflic_!!!.???", "aflic_" + strings.Repeat("a", 10) + ".short",
		// A payload that decodes but is not JSON.
		"aflic_" + base64.RawURLEncoding.EncodeToString([]byte("hello")) + "." +
			base64.RawURLEncoding.EncodeToString(make([]byte, ed25519.SignatureSize)),
	} {
		_, err := v.Parse(token)
		require.Errorf(t, err, "token %q was accepted", token)
	}
}

func TestParse_RefusesClaimsMissingWhatMakesALicenseMeanAnything(t *testing.T) {
	t.Parallel()
	s := newSigner(t, "k")
	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})

	noOrg := validClaims()
	noOrg.Org = ""
	_, err := v.Parse(s.sign(t, noOrg))
	// A license naming no organization would work everywhere.
	require.ErrorIs(t, err, license.ErrMalformed)

	noExpiry := validClaims()
	noExpiry.ExpiresAt = time.Time{}
	_, err = v.Parse(s.sign(t, noExpiry))
	// One with no expiry never ends.
	require.ErrorIs(t, err, license.ErrMalformed)
}

// FuzzParse is the parser's real test. It reads attacker-controlled input, and
// the only property that matters is that no input makes it panic or accept
// something it did not sign.
func FuzzParse(f *testing.F) {
	pub, priv, err := ed25519.GenerateKey(nil)
	if err != nil {
		f.Fatal(err)
	}
	v := license.NewVerifier(map[string]ed25519.PublicKey{"k": pub})

	payload, _ := json.Marshal(license.Claims{
		ID: "l", Org: "acme", ExpiresAt: epoch, KeyID: "k",
	})
	good := fmt.Sprintf("aflic_%s.%s",
		base64.RawURLEncoding.EncodeToString(payload),
		base64.RawURLEncoding.EncodeToString(ed25519.Sign(priv, payload)))

	for _, seed := range []string{good, "", "aflic_", "aflic_a.b", "\x00", strings.Repeat("a", 4096)} {
		f.Add(seed)
	}

	f.Fuzz(func(t *testing.T, token string) {
		claims, err := v.Parse(token)
		if err != nil {
			return
		}
		// Anything accepted must have been signed by the key, and must carry the
		// fields that make a license mean something.
		if claims.Org == "" || claims.ExpiresAt.IsZero() {
			t.Fatalf("accepted a license with no organization or no expiry: %+v", claims)
		}
		if claims.KeyID != "k" {
			t.Fatalf("accepted a license signed by %q", claims.KeyID)
		}
	})
}

// ---------------------------------------------------------------------------
// Evaluation
// ---------------------------------------------------------------------------

func evaluate(t *testing.T, claims license.Claims, ev license.Evaluation) license.Status {
	t.Helper()
	s := newSigner(t, "k")
	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})
	parsed, err := v.Parse(s.sign(t, claims))
	require.NoError(t, err)
	return v.Evaluate(parsed, ev)
}

func TestEvaluate_ACurrentLicenseGrantsExactlyWhatItNames(t *testing.T) {
	t.Parallel()
	status := evaluate(t, validClaims(), license.Evaluation{Org: "acme", Now: epoch.AddDate(0, 1, 0)})

	require.Equal(t, license.StateActive, status.State)
	require.True(t, status.Honoured())
	require.True(t, status.Enabled(license.FeatureSSO))
	require.True(t, status.Enabled(license.FeatureSCIM))
	// And nothing it does not name. A license that grants everything it was not
	// sold is a license nobody needs to buy the right tier of.
	require.False(t, status.Enabled(license.FeatureBilling))
	require.False(t, status.Enabled(license.FeatureMultiRuntime))
}

func TestEvaluate_TheEntitlementMatrix(t *testing.T) {
	t.Parallel()
	// Every feature, granted and not granted, so that a feature added to the
	// list without being handled fails here.
	for _, granted := range license.AllFeatures() {
		claims := validClaims()
		claims.Features = []license.Feature{granted}
		status := evaluate(t, claims, license.Evaluation{Org: "acme", Now: epoch.AddDate(0, 1, 0)})

		for _, f := range license.AllFeatures() {
			if f == granted {
				require.Truef(t, status.Enabled(f), "a license naming %s does not grant it", f)
				continue
			}
			require.Falsef(t, status.Enabled(f), "a license naming only %s also granted %s", granted, f)
		}
	}
}

func TestEvaluate_ALicenseForAnotherOrganizationIsRefused(t *testing.T) {
	t.Parallel()
	// What stops a key from being passed between companies.
	status := evaluate(t, validClaims(), license.Evaluation{Org: "other-co", Now: epoch})

	require.Equal(t, license.StateWrongOrg, status.State)
	require.False(t, status.Honoured())
	for _, f := range license.AllFeatures() {
		require.Falsef(t, status.Enabled(f), "%s was granted to the wrong organization", f)
	}
	require.Contains(t, status.Warning, "acme")
	require.Contains(t, status.Warning, "other-co")
}

func TestEvaluate_TheOrganizationComparisonIgnoresCaseAndSurroundingSpace(t *testing.T) {
	t.Parallel()
	status := evaluate(t, validClaims(), license.Evaluation{Org: "  ACME ", Now: epoch})
	require.Equal(t, license.StateActive, status.State)
}

func TestEvaluate_ExpiryMovesThroughGraceAndThenDegrades(t *testing.T) {
	t.Parallel()
	claims := validClaims()
	claims.GraceDays = 14
	expires := claims.ExpiresAt

	for _, tc := range []struct {
		name  string
		now   time.Time
		state license.State
		sso   bool
	}{
		{"a month before", expires.AddDate(0, 0, -30), license.StateActive, true},
		{"the day before", expires.AddDate(0, 0, -1), license.StateActive, true},
		{"the day after", expires.AddDate(0, 0, 1), license.StateGrace, true},
		{"the last day of grace", expires.AddDate(0, 0, 13), license.StateGrace, true},
		{"the day grace ends", expires.AddDate(0, 0, 14), license.StateExpired, false},
		{"a year later", expires.AddDate(1, 0, 0), license.StateExpired, false},
	} {
		t.Run(tc.name, func(t *testing.T) {
			status := evaluate(t, claims, license.Evaluation{Org: "acme", Now: tc.now})
			require.Equal(t, tc.state, status.State)
			require.Equal(t, tc.sso, status.Enabled(license.FeatureSSO))
		})
	}
}

func TestEvaluate_AnExpiredLicenseSaysNothingIsDeleted(t *testing.T) {
	t.Parallel()
	// The behaviour that matters more than the mechanism: settings survive
	// expiry, and renewing restores them. Saying so is what stops somebody from
	// reconfiguring everything after a late renewal.
	claims := validClaims()
	status := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: claims.ExpiresAt.AddDate(0, 1, 0),
	})
	require.Equal(t, license.StateExpired, status.State)
	require.Contains(t, status.Warning, "preserved")
	require.Contains(t, status.Warning, "renewing")
}

func TestEvaluate_WarnsBeforeExpiryRatherThanAtIt(t *testing.T) {
	t.Parallel()
	claims := validClaims()
	// Thirty days out, which is enough time for a purchase order.
	status := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: claims.ExpiresAt.AddDate(0, 0, -20),
	})
	require.Equal(t, license.StateActive, status.State)
	require.Contains(t, status.Warning, "expires in 20 days")

	quiet := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: claims.ExpiresAt.AddDate(0, 0, -200),
	})
	require.Empty(t, quiet.Warning, "a license with months left should say nothing")
}

func TestEvaluate_ClockRollbackDegradesUntilTheClockCatchesUp(t *testing.T) {
	t.Parallel()
	claims := validClaims()
	lastSeen := epoch.AddDate(0, 6, 0)

	// The clock moved back six months, which is how an expired license is made
	// to look current.
	rolled := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: epoch.AddDate(0, 0, 1), LastSeen: lastSeen,
	})
	require.Equal(t, license.StateClockRollback, rolled.State)
	require.False(t, rolled.Enabled(license.FeatureSSO))
	require.Contains(t, rolled.Warning, "clock")

	// Back past the last seen time and it works again, with nothing to undo.
	recovered := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: lastSeen.Add(time.Second), LastSeen: lastSeen,
	})
	require.Equal(t, license.StateActive, recovered.State)
	require.True(t, recovered.Enabled(license.FeatureSSO))
}

func TestEvaluate_OrdinaryClockCorrectionIsNotTreatedAsRollback(t *testing.T) {
	t.Parallel()
	// Time synchronisation moves a clock by seconds and a resumed virtual
	// machine by more. Treating either as tampering would disable enterprise
	// features on a laptop that woke from sleep.
	claims := validClaims()
	lastSeen := epoch.AddDate(0, 6, 0)

	for _, back := range []time.Duration{time.Second, time.Minute, 59 * time.Minute} {
		status := evaluate(t, claims, license.Evaluation{
			Org: "acme", Now: lastSeen.Add(-back), LastSeen: lastSeen,
		})
		require.Equalf(t, license.StateActive, status.State,
			"a clock %s behind was treated as tampering", back)
	}

	// An hour and a bit is past the tolerance.
	beyond := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: lastSeen.Add(-2 * time.Hour), LastSeen: lastSeen,
	})
	require.Equal(t, license.StateClockRollback, beyond.State)
}

func TestEvaluate_TheClockCheckHappensBeforeExpiry(t *testing.T) {
	t.Parallel()
	// The order is the point. An expired license with the clock wound back looks
	// current, so checking expiry first would let exactly the attack this
	// prevents succeed.
	claims := validClaims()
	lastSeen := claims.ExpiresAt.AddDate(0, 1, 0) // already seen after expiry

	status := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: epoch.AddDate(0, 1, 0), LastSeen: lastSeen,
	})
	require.Equal(t, license.StateClockRollback, status.State)
	require.False(t, status.Enabled(license.FeatureSSO))
}

func TestEvaluate_ARevokedLicenseIsRefusedEvenWhileCurrent(t *testing.T) {
	t.Parallel()
	s := newSigner(t, "k")
	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})
	claims, err := v.Parse(s.sign(t, validClaims()))
	require.NoError(t, err)

	before := v.Evaluate(claims, license.Evaluation{Org: "acme", Now: epoch})
	require.Equal(t, license.StateActive, before.State)

	v.Revoke("lic-1")
	after := v.Evaluate(claims, license.Evaluation{Org: "acme", Now: epoch})
	require.Equal(t, license.StateRevoked, after.State)
	require.False(t, after.Enabled(license.FeatureSSO))
}

func TestEvaluate_ATrialSaysSo(t *testing.T) {
	t.Parallel()
	claims := validClaims()
	claims.Trial = true
	status := evaluate(t, claims, license.Evaluation{Org: "acme", Now: epoch})
	require.Contains(t, status.Warning, "trial")
	require.True(t, status.Enabled(license.FeatureSSO))
}

// ---------------------------------------------------------------------------
// No license, which is the ordinary case
// ---------------------------------------------------------------------------

func TestNone_IsTheCommunityEditionAndNotAnError(t *testing.T) {
	t.Parallel()
	status := license.None()
	require.Equal(t, license.StateNone, status.State)
	require.False(t, status.Honoured())
	require.Empty(t, status.Warning,
		"most installations have no license, deliberately and permanently; warning them would be untrue")
	for _, f := range license.AllFeatures() {
		require.False(t, status.Enabled(f))
	}
}

func TestZeroStatus_GrantsNothing(t *testing.T) {
	t.Parallel()
	// A Status that was never evaluated must not grant anything. The direction
	// the mistake fails in is the whole design.
	var zero license.Status
	for _, f := range license.AllFeatures() {
		require.Falsef(t, zero.Enabled(f), "the zero Status granted %s", f)
	}
	require.False(t, zero.Honoured())
}

// ---------------------------------------------------------------------------
// Seats
// ---------------------------------------------------------------------------

func TestSeats_AreReportedAndNeverEnforcedByRemovingAnyone(t *testing.T) {
	t.Parallel()
	claims := validClaims()
	claims.Seats = 3
	status := evaluate(t, claims, license.Evaluation{Org: "acme", Now: epoch})

	require.False(t, status.SeatsExceeded(2), "there is room for a third")
	require.True(t, status.SeatsExceeded(3), "a fourth is past the limit")
	require.True(t, status.SeatsExceeded(99))
}

func TestSeats_ZeroMeansUnlimited(t *testing.T) {
	t.Parallel()
	claims := validClaims()
	claims.Seats = 0
	status := evaluate(t, claims, license.Evaluation{Org: "acme", Now: epoch})
	require.False(t, status.SeatsExceeded(100_000))
}

func TestSeats_AreNotEnforcedWithoutAnHonouredLicense(t *testing.T) {
	t.Parallel()
	// An expired license must not start refusing new members. Degrading to the
	// community behaviour means the community behaviour, which has no seats.
	claims := validClaims()
	claims.Seats = 1
	status := evaluate(t, claims, license.Evaluation{
		Org: "acme", Now: claims.ExpiresAt.AddDate(1, 0, 0),
	})
	require.Equal(t, license.StateExpired, status.State)
	require.False(t, status.SeatsExceeded(500))
}

// ---------------------------------------------------------------------------

func TestAllFeatures_IsSortedAndComplete(t *testing.T) {
	t.Parallel()
	all := license.AllFeatures()
	require.NotEmpty(t, all)
	for i := 1; i < len(all); i++ {
		require.Less(t, string(all[i-1]), string(all[i]), "AllFeatures must be sorted")
	}
	seen := map[license.Feature]bool{}
	for _, f := range all {
		require.False(t, seen[f], "%s appears twice", f)
		seen[f] = true
	}
}

func TestErrors_AreDistinguishable(t *testing.T) {
	t.Parallel()
	// A caller maps each to a different error code and a different next step:
	// a malformed key is a paste error, a tampered one is a security event, and
	// an unknown key means the build is older than the license.
	require.False(t, errors.Is(license.ErrMalformed, license.ErrTampered))
	require.False(t, errors.Is(license.ErrTampered, license.ErrUnknownKey))
	require.False(t, errors.Is(license.ErrUnknownKey, license.ErrMalformed))
}

// TestParse_ToleratesAFeatureThisBuildDoesNotKnow pins the permissiveness.
//
// It looks like a bug and it is load bearing. A license issued for a newer
// release names features an older binary does not carry, and a verifier that
// refused the whole license over one unknown name would take away the features
// the customer did buy every time an upgrade and a renewal crossed. The unknown
// name is carried and never permitted, and tools/licensegen is what stops one
// being signed in the first place.
func TestParse_ToleratesAFeatureThisBuildDoesNotKnow(t *testing.T) {
	s := newSigner(t, "k1")
	claims := validClaims()
	// One real feature and one from a release this build predates.
	claims.Features = []license.Feature{license.FeatureSSO, license.Feature("a_later_release")}

	v := license.NewVerifier(map[string]ed25519.PublicKey{s.keyID: s.pub})
	parsed, err := v.Parse(s.sign(t, claims))
	require.NoError(t, err, "an unknown feature must not invalidate the whole license")

	status := v.Evaluate(parsed, license.Evaluation{Org: claims.Org, Now: claims.IssuedAt})
	require.Equal(t, license.StateActive, status.State)
	require.True(t, status.Enabled(license.FeatureSSO), "the known feature is still permitted")
	// The unknown name is carried through and would answer true if anything
	// asked, which nothing does: every caller asks by constant, and this build
	// has no constant for a feature a later release introduced. Asserted rather
	// than left implied, because the harmless version of this and the dangerous
	// version look identical from the outside.
	require.True(t, status.Enabled(license.Feature("a_later_release")),
		"the unknown name is carried rather than dropped")
	for _, f := range license.AllFeatures() {
		if f == license.FeatureSSO {
			continue
		}
		require.False(t, status.Enabled(f), "no feature this build knows is granted by accident")
	}
}

File: ee/web/server/test/license.test.ts
// The licence reader, and the seam that must not be half-installed.
//
// Not MIT. Covered by the Antifailure Enterprise License; see ee/LICENSE.md.
//
// entrypoint.test.ts proves the gate over a socket, which is the claim that
// matters, and it can only afford a handful of cases because each one is a
// process. This file is the cheap half: the refusals a token can produce, and
// the states a valid one moves through, driven directly.

import { describe, it } from 'node:test'
import assert from 'node:assert/strict'
import { generateKeyPairSync, randomUUID, sign } from 'node:crypto'
import { clearExtensions, hasSignInPolicy, registeredExtensions, setSignInPolicy, FakeClock } from '@antifailure/api'
import {
  LicenseRefused,
  evaluate,
  licenseFromEnv,
  none,
  parseLicense,
  trustedKeys,
} from '../src/license.ts'
import { registerEnterprise } from '../src/register.ts'

function key(): { kid: string; publicKeys: string; token: (claims: object) => string; other: string } {
  const kid = `k-${randomUUID().slice(0, 8)}`
  const { publicKey, privateKey } = generateKeyPairSync('ed25519')
  const raw = publicKey.export({ format: 'der', type: 'spki' }).subarray(12)
  const impostor = generateKeyPairSync('ed25519')
  return {
    kid,
    publicKeys: `${kid}=${raw.toString('base64')}`,
    token: (claims) => {
      const payload = Buffer.from(JSON.stringify({ kid, ...claims }), 'utf8')
      return `aflic_${payload.toString('base64url')}.${sign(null, payload, privateKey).toString('base64url')}`
    },
    // A token signed by a key that is real and is not ours, which is the shape
    // of somebody minting their own licences rather than of a corrupted one.
    other: (() => {
      const payload = Buffer.from(JSON.stringify({ kid, org: 'acme', expires_at: '2030-01-01T00:00:00Z' }), 'utf8')
      return `aflic_${payload.toString('base64url')}.${sign(null, payload, impostor.privateKey).toString('base64url')}`
    })(),
  }
}

const future = '2030-01-01T00:00:00Z'

describe('parsing a license key', () => {
  it('accepts one this installation trusts', () => {
    const k = key()
    const claims = parseLicense(k.token({ org: 'acme', features: ['sso'], expires_at: future }), trustedKeys(k.publicKeys))
    assert.equal(claims.org, 'acme')
    assert.deepEqual(claims.features, ['sso'])
  })

  it('accepts one pasted out of an email with newlines in it', () => {
    // Refusing whitespace is a support ticket rather than a security property,
    // and the Go side tolerates it for the same reason.
    const k = key()
    const token = k.token({ org: 'acme', expires_at: future })
    const wrapped = `${token.slice(0, 20)}\n  ${token.slice(20)}\n`
    assert.equal(parseLicense(wrapped, trustedKeys(k.publicKeys)).org, 'acme')
  })

  it('refuses a signature that does not verify', () => {
    const k = key()
    assert.throws(
      () => parseLicense(k.other, trustedKeys(k.publicKeys)),
      (err: unknown) => err instanceof LicenseRefused && err.refusal === 'tampered',
    )
  })

  it('refuses a payload edited after signing', () => {
    // The one that matters commercially: adding a feature to a licence you
    // already hold. The claims are re-serialised, so the signature is over
    // different bytes.
    const k = key()
    const token = k.token({ org: 'acme', features: ['sso'], expires_at: future })
    const [body, signature] = token.slice('aflic_'.length).split('.')
    const edited = JSON.parse(Buffer.from(body!, 'base64url').toString('utf8')) as Record<string, unknown>
    edited.features = ['sso', 'scim', 'air_gapped']
    const forged = `aflic_${Buffer.from(JSON.stringify(edited), 'utf8').toString('base64url')}.${signature}`
    assert.throws(
      () => parseLicense(forged, trustedKeys(k.publicKeys)),
      (err: unknown) => err instanceof LicenseRefused && err.refusal === 'tampered',
    )
  })

  it('refuses a key this installation does not carry, and says that rather than tampered', () => {
    // Two different problems and two different places to send somebody. A build
    // with no keys told to look for tampering is a support ticket that ends in
    // embarrassment.
    const k = key()
    assert.throws(
      () => parseLicense(k.token({ org: 'acme', expires_at: future }), trustedKeys(undefined)),
      (err: unknown) =>
        err instanceof LicenseRefused && err.refusal === 'unknown_key' && /carries no licence signing keys/.test(err.message),
    )
  })

  it('refuses a token that is not a license at all', () => {
    const k = key()
    const keys = trustedKeys(k.publicKeys)
    for (const bad of ['', 'hello', 'aflic_', 'aflic_abc', 'aflic_.abc', 'aflic_abc.', 'aflic_!!!.!!!']) {
      assert.throws(
        () => parseLicense(bad, keys),
        (err: unknown) => err instanceof LicenseRefused && err.refusal === 'malformed',
        `${JSON.stringify(bad)} was not refused as malformed`,
      )
    }
  })

  it('refuses one that names no organization, so a key cannot be passed around', () => {
    const k = key()
    assert.throws(
      () => parseLicense(k.token({ expires_at: future }), trustedKeys(k.publicKeys)),
      (err: unknown) => err instanceof LicenseRefused && /names no organization/.test(err.message),
    )
  })

  it('carries a feature name it has never heard of without refusing the license', () => {
    // The set is closed at issue time, not here. A licence issued for a newer
    // release names features an older build has not heard of, and refusing the
    // whole licence over one would turn every ordering of upgrade and renewal
    // into an outage for features the customer did buy.
    //
    // The unknown name lands in the permitted set, exactly as it does in the Go
    // verifier, whose loop is `for _, f := range claims.Features`. That is
    // parity rather than an oversight, and it is deliberately not "improved"
    // here: two implementations of one decision that differ in a corner nobody
    // observes is how they start differing in one somebody does. Nothing can
    // ask, because Feature is a closed union on this side and a closed const
    // set on that one, so the only reachable answers are the ones below.
    //
    // Worth writing down, because ee/engine/license's own doc comment says the
    // verifier "simply never permits" an unknown name, and both implementations
    // would permit it if anything could ask. Unobservable today, and untrue as
    // written.
    const k = key()
    const claims = parseLicense(
      k.token({ org: 'acme', features: ['sso', 'time_travel'], expires_at: future }),
      trustedKeys(k.publicKeys),
    )
    const status = evaluate(claims, { org: 'acme', now: new Date('2027-01-01T00:00:00Z') })
    assert.equal(status.enabled('sso'), true)
    assert.equal(status.enabled('scim'), false)
    assert.deepEqual(claims.features, ['sso', 'time_travel'])
  })
})

describe('what a license is doing right now', () => {
  const k = key()
  const claims = (over: Record<string, unknown> = {}) =>
    parseLicense(
      k.token({ id: 'lic-1', org: 'acme', features: ['sso', 'scim'], expires_at: future, ...over }),
      trustedKeys(k.publicKeys),
    )

  it('is active before expiry and permits what it names', () => {
    const status = evaluate(claims(), { org: 'acme', now: new Date('2029-01-01T00:00:00Z') })
    assert.equal(status.state, 'active')
    assert.equal(status.enabled('sso'), true)
    assert.equal(status.honoured(), true)
  })

  it('keeps working through the grace period rather than stopping', () => {
    // Nobody's sign-in breaks because a purchase order moved slowly.
    const status = evaluate(claims(), { org: 'acme', now: new Date('2030-01-05T00:00:00Z') })
    assert.equal(status.state, 'grace')
    assert.equal(status.enabled('sso'), true)
    // Fourteen days of grace from 1 January, read on the fifth, so ten remain.
    assert.match(status.warning, /keep working for 10 more days/)
  })

  it('permits nothing once the grace period has ended', () => {
    const status = evaluate(claims(), { org: 'acme', now: new Date('2030-02-01T00:00:00Z') })
    assert.equal(status.state, 'expired')
    assert.equal(status.enabled('sso'), false)
    assert.match(status.warning, /every enterprise setting is preserved/)
  })

  it('refuses a license issued to somebody else', () => {
    const status = evaluate(claims(), { org: 'other', now: new Date('2029-01-01T00:00:00Z') })
    assert.equal(status.state, 'wrong_org')
    assert.equal(status.enabled('sso'), false)
  })

  it('refuses when the clock has moved backwards past a time it was seen at', () => {
    // Offline verification is only as good as the clock, and moving it back is
    // how an expired licence is made to look current.
    const status = evaluate(claims(), {
      org: 'acme',
      now: new Date('2029-01-01T00:00:00Z'),
      lastSeen: new Date('2029-06-01T00:00:00Z'),
    })
    assert.equal(status.state, 'clock_rollback')
    assert.equal(status.enabled('sso'), false)
  })

  it('tolerates an hour of clock movement, because a resumed virtual machine is not an attack', () => {
    const status = evaluate(claims(), {
      org: 'acme',
      now: new Date('2029-01-01T00:00:00Z'),
      lastSeen: new Date('2029-01-01T00:30:00Z'),
    })
    assert.equal(status.state, 'active')
  })

  it('refuses a revoked license without asking anything on the network', () => {
    const status = evaluate(claims(), {
      org: 'acme',
      now: new Date('2029-01-01T00:00:00Z'),
      revoked: new Set(['lic-1']),
    })
    assert.equal(status.state, 'revoked')
  })

  it('reports the seat limit rather than enforcing it by removing anybody', () => {
    const status = evaluate(claims({ seats: 3 }), { org: 'acme', now: new Date('2029-01-01T00:00:00Z') })
    assert.equal(status.seatsExceeded(2), false)
    assert.equal(status.seatsExceeded(3), true)
    // Zero seats is unlimited, not a limit of nobody.
    const unlimited = evaluate(claims({ seats: 0 }), { org: 'acme', now: new Date('2029-01-01T00:00:00Z') })
    assert.equal(unlimited.seatsExceeded(10_000), false)
  })

  it('permits nothing at all with no license, and calls that none rather than expired', () => {
    const status = none()
    assert.equal(status.state, 'none')
    assert.equal(status.enabled('sso'), false)
    assert.equal(status.honoured(), false)
  })

  it('refuses a license key set with no organization to check it against', () => {
    assert.throws(
      () => licenseFromEnv({ AF_LICENSE_KEY: k.token({ org: 'acme', expires_at: future }) }, new Date()),
      (err: unknown) => err instanceof LicenseRefused && /AF_ORG is not/.test(err.message),
    )
  })
})

describe('registering the enterprise edition', () => {
  it('installs both sign-on extension points, never one', () => {
    // ee/web/sso/src/index.ts says why at length: an organization that has
    // REQUIRED single sign-on still has GitHub sign-in open unless the policy
    // is installed too, which is a feature that looks complete and enforces
    // nothing. This caller is the one that did not exist, so nothing checked
    // that it does both.
    clearExtensions()
    setSignInPolicy(null)
    assert.equal(hasSignInPolicy(), false)

    registerEnterprise({
      pool: {} as never,
      clock: new FakeClock(),
      baseUrl: 'https://enterprise.test',
      appBaseUrl: 'https://enterprise.test/',
      secureCookies: true,
      env: {},
      log: () => {},
      encryptionKey: Buffer.alloc(32, 7),
    })

    assert.deepEqual(registeredExtensions().map((e) => e.name).sort(), ['scim', 'sso'])
    assert.equal(
      hasSignInPolicy(),
      true,
      'the routes were registered and the sign-in policy was not, so an organization that ' +
        'requires single sign-on would still be reachable through GitHub',
    )
    clearExtensions()
    setSignInPolicy(null)
  })
})
